# Exploring Ebay Car Sales Data

In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

https://www.kaggle.com/orgesleka/used-cars-database/data

The aim of this project is to clean the data and analyze the included used car listings.

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding='Latin-1')

autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

50000 rows, most are strings, some are integers. A few categories with null values.

## Cleaning Column Names

Let's convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
autos = autos.copy()
autos.rename(columns = {'yearOfRegistration': 'registration_year', 
                        'monthOfRegistration': 'registration_month', 
                        'notRepairedDamage': 'unrepaired_damage',
                        'dateCreated': 'ad_created',
                        'dateCrawled': 'date_crawled',
                        'offerType': 'offer_type',
                        'vehicleType': 'vehicle_type',
                        'powerPS': 'power_PS',
                        'fuelType': 'fuel_type',
                        'nrOfPictures': 'nr_of_pictures',
                        'postalCode': 'postal_code',
                        'lastSeen': 'last_seen'
                        }, inplace=True)


In [8]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Changed the column names from camelcase to snakecase, and some of the column names were changed to readability.

## Initial Exploration and Cleaning

Let's do some basic data exploration to determine what other cleaning tasks need to be done. Initially we will look for: 
- Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis. 
- Examples of numeric data stored as text which can be cleaned and converted.

In [9]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-04 16:40:33  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Looking at the description of the columns, there are candidates for drops (seller, offer_type, abtest, unrepaired_damage, nr_of_pictures).

Price and odometer are stored as text, should be converted to numeric.

In [10]:
autos["price"] = (autos["price"]
                  .str.replace('$', '')
                  .str.replace(",", "")
                  .astype(int)
                 )

autos["odometer"] = (autos["odometer"]
                    .str.replace('km', '')
                    .str.replace(",", "")
                    .astype(int)
                    )
                    
autos.rename(columns = {"odometer": "odometer_km"}, 
            inplace=True
            )

In [11]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_PS   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [12]:
autos["price"].unique().shape

(2357,)

In [13]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [14]:
autos["price"].value_counts().sort_index(ascending=True).head()

0    1421
1     156
2       3
3       1
5       2
Name: price, dtype: int64

## Exploring the Odometer and Price Columns

Let's continue exploring the data, specifically looking for data that doesn't look right. We'll start by analyzing the odometer_km and price columns.

In [15]:
autos["odometer_km"].unique().shape

(13,)

In [16]:
autos["odometer_km"].value_counts().sort_index(ascending=True).head()

5000     967
10000    264
20000    784
30000    789
40000    819
Name: odometer_km, dtype: int64

Looks like the price column has an outlier or two, let's check it out some more:

In [17]:
autos.nlargest(10, "price")

date_crawled                                               name  \
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
47634  2016-04-04 21:25:21                                        Ferrari_FXX   
7814   2016-04-04 11:53:31                                        Ferrari_F40   

       seller offer_type     price   abtest vehicle_type  registration_year  \
39705  privat    Angebot  99999999  control    limousine               1999   
42221  privat    Angebot  27322222  control    limousine               2014   
27371  privat    Angebot  12345678  control          NaN               2017   
39377  privat    Angebot  12345678  control          NaN               2018   
47598  privat    Angebot  12345678  control    limousine               2001   
2897   privat    Angebot  11111111     test    limousine               1973   
24384  privat    Angebot  11111111     test          NaN               1995   
11137  privat    Angebot  10000000  control        coupe               1960   
47634  privat    Angebot   3890000     test        coupe               2006   
7814   privat    Angebot   1300000  control        coupe               1992   

         gearbox  power_PS     model  odometer_km  registration_month  \
39705  automatik       224  s_klasse       150000                   9   
42221    manuell       163        c4        40000                   2   
27371        NaN        95     punto       150000                   0   
39377    manuell        95       v40       150000                   6   
47598    manuell       101    vectra       150000                   3   
2897     manuell        48    escort        50000                   3   
24384        NaN         0       NaN       150000                   0   
11137    manuell       368       NaN       100000                   1   
47634        NaN       799       NaN         5000                   7   
7814         NaN         0       NaN        50000                  12   

      fuel_type           brand unrepaired_damage           ad_created  \
39705    benzin   mercedes_benz               NaN  2016-03-22 00:00:00   
42221    diesel         citroen               NaN  2016-03-08 00:00:00   
27371       NaN            fiat               NaN  2016-03-09 00:00:00   
39377       NaN           volvo              nein  2016-03-08 00:00:00   
47598    benzin            opel              nein  2016-03-31 00:00:00   
2897     benzin            ford              nein  2016-03-12 00:00:00   
24384       NaN      volkswagen               NaN  2016-03-21 00:00:00   
11137    benzin  sonstige_autos              nein  2016-03-29 00:00:00   
47634       NaN  sonstige_autos              nein  2016-04-04 00:00:00   
7814        NaN  sonstige_autos              nein  2016-04-04 00:00:00   

       nr_of_pictures  postal_code            last_seen  
39705               0        73525  2016-04-06 05:15:30  
42221               0        76532  2016-03-08 20:39:05  
27371               0        96110  2016-03-09 15:45:47  
39377               0        14542  2016-04-06 23:17:31  
47598               0         4356  2016-03-31 18:56:54  
2897                0        94469  2016-03-12 22:45:27  
24384               0        18519  2016-03-21 14:40:18  
11137               0        73033  2016-04-06 21:18:11  
47634               0        60313  2016-04-05 12:07:37  
7814                0        60598  2016-04-05 11:34:

The Ferrari FXX is around a \$4 million sports car, but the maserati 3200 gt is not \$10 million. We will set the max price at \$4 million to keep the Ferrari.

There are 1421 rows with 0, which is about 2% of all values, so we can remove those too. There are 156 rows with \$1, but there are such things as \$1 starting bids for ebay auctions, so we will keep those.

In [18]:
autos = autos[autos["price"].between(1,4000000)]

In [19]:
autos.describe(include='all')

date_crawled         name  seller offer_type         price  \
count                 48571        48571   48571      48571  4.857100e+04   
unique                46888        37475       2          1           NaN   
top     2016-03-12 16:06:22  Ford_Fiesta  privat    Angebot           NaN   
freq                      3           76   48570      48571           NaN   
mean                    NaN          NaN     NaN        NaN  6.082245e+03   
std                     NaN          NaN     NaN        NaN  2.278698e+04   
min                     NaN          NaN     NaN        NaN  1.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.200000e+03   
50%                     NaN          NaN     NaN        NaN  3.000000e+03   
75%                     NaN          NaN     NaN        NaN  7.490000e+03   
max                     NaN          NaN     NaN        NaN  3.890000e+06   

       abtest vehicle_type  registration_year  gearbox      power_PS  model  \
count   48571        43985       48571.000000    46226  48571.000000  46110   
unique      2            8                NaN        2           NaN    245   
top      test    limousine                NaN  manuell           NaN   golf   
freq    25023        12599                NaN    36105           NaN   3900   
mean      NaN          NaN        2004.754257      NaN    117.208849    NaN   
std       NaN          NaN          88.638548      NaN    200.662179    NaN   
min       NaN          NaN        1000.000000      NaN      0.000000    NaN   
25%       NaN          NaN        1999.000000      NaN     71.000000    NaN   
50%       NaN          NaN        2004.000000      NaN    107.000000    NaN   
75%       NaN          NaN        2008.000000      NaN    150.000000    NaN   
max       NaN          NaN        9999.000000      NaN  17700.000000    NaN   

          odometer_km  registration_month fuel_type       brand  \
count    48571.000000        48571.000000     44539       48571   
unique            NaN                 NaN         7          40   
top               NaN                 NaN    benzin  volkswagen   
freq              NaN                 NaN     29371       10338   
mean    125767.536184            5.782319       NaN         NaN   
std      39791.893858            3.685700       NaN         NaN   
min       5000.000000            0.000000       NaN         NaN   
25%     125000.000000            3.000000       NaN         NaN   
50%     150000.000000            6.000000       NaN         NaN   
75%     150000.000000            9.000000       NaN         NaN   
max     150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count              39468                48571         48571.0  48571.000000   
unique                 2                   76             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq               34778                 1887             NaN           NaN   
mean                 NaN                  NaN             0.0  50975.654465   
std                  NaN                  NaN             0.0  25747.456521   
min                  NaN                  NaN             0.0   1067.000000   
25%                  NaN                  NaN             0.0  30657.000000   
50%                  NaN                  NaN             0.0  49716.000000   
75%                  NaN                  NaN             0.0  71665.000000   
max                  NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 48571  
unique                38479  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

## Exploring the Date Columns

In [20]:
autos['date_crawled'] = autos['date_crawled'].str[:10]

In [21]:
autos['ad_created'] = autos['ad_created'].str[:10]

In [22]:
autos['last_seen'] = autos['last_seen'].str[:10]

In [23]:
autos["date_crawled"].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025324
2016-03-06    0.014041
2016-03-07    0.036009
2016-03-08    0.033291
2016-03-09    0.033086
2016-03-10    0.032180
2016-03-11    0.032571
2016-03-12    0.036915
2016-03-13    0.015668
2016-03-14    0.036544
2016-03-15    0.034280
2016-03-16    0.029606
2016-03-17    0.031644
2016-03-18    0.012909
2016-03-19    0.034774
2016-03-20    0.037883
2016-03-21    0.037389
2016-03-22    0.033003
2016-03-23    0.032221
2016-03-24    0.029338
2016-03-25    0.031603
2016-03-26    0.032200
2016-03-27    0.031089
2016-03-28    0.034856
2016-03-29    0.034115
2016-03-30    0.033683
2016-03-31    0.031830
2016-04-01    0.033683
2016-04-02    0.035474
2016-04-03    0.038603
2016-04-04    0.036524
2016-04-05    0.013094
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

The site was crawled for about a month between March and April. Uniform distribution of number of listing crawled.

In [24]:
autos["ad_created"].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033147
2016-03-10    0.031891
2016-03-11    0.032900
2016-03-12    0.036750
2016-03-13    0.017006
2016-03-14    0.035186
2016-03-15    0.034012
2016-03-16    0.030121
2016-03-17    0.031294
2016-03-18    0.013588
2016-03-19    0.033683
2016-03-20    0.037944
2016-03-21 

There are some ads that were created 10 months ago! Most ads were created quite recently.

In [25]:
autos["last_seen"].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005394
2016-03-08    0.007412
2016-03-09    0.009594
2016-03-10    0.010665
2016-03-11    0.012374
2016-03-12    0.023780
2016-03-13    0.008894
2016-03-14    0.012600
2016-03-15    0.015874
2016-03-16    0.016450
2016-03-17    0.028083
2016-03-18    0.007350
2016-03-19    0.015832
2016-03-20    0.020650
2016-03-21    0.020630
2016-03-22    0.021371
2016-03-23    0.018530
2016-03-24    0.019765
2016-03-25    0.019209
2016-03-26    0.016800
2016-03-27    0.015647
2016-03-28    0.020877
2016-03-29    0.022359
2016-03-30    0.024768
2016-03-31    0.023780
2016-04-01    0.022791
2016-04-02    0.024933
2016-04-03    0.025200
2016-04-04    0.024480
2016-04-05    0.124786
2016-04-06    0.221799
2016-04-07    0.131931
Name: last_seen, dtype: float64

The last_seen column shows the last date an ad was seen by the crawler, presumably due to the vehicle being sold. It is quite even until the last 3 days! This is probably becuase the crawler was shutting down at the end, and therefore is not indicative of when the car was sold.

In [26]:
autos['registration_year'].describe()

count    48571.000000
mean      2004.754257
std         88.638548
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

## Dealing with Incorrect Registration Year Data

We have some interesting numbers for min and max: min of 1000 was made many hundreds of years before the model t first rolled down the factory, and max of 9999 is many thousands of years into the future. We'll keep up to 2016, and we have some vehicles < 1911, so we'll take those out

In [27]:
autos['registration_year'].nsmallest(25)
                                    

22316    1000
49283    1001
24511    1111
10556    1800
32585    1800
3679     1910
22659    1910
28693    1910
30781    1910
45157    1910
21416    1927
22101    1929
11246    1931
2221     1934
2573     1934
21421    1937
23804    1937
26607    1937
39725    1937
26103    1938
24855    1939
13963    1941
25792    1941
11585    1943
11047    1948
Name: registration_year, dtype: int64

In [28]:
autos = autos[autos["registration_year"].between(1911,2016)]

In [29]:
autos['registration_year'].value_counts(normalize=True)

2000    0.067606
2005    0.062894
1999    0.062080
2004    0.057902
2003    0.057817
2006    0.057217
2001    0.056467
2002    0.053254
1998    0.050619
2007    0.048777
2008    0.047449
2009    0.044685
1997    0.041793
2011    0.034767
2010    0.034039
1996    0.029412
2012    0.028062
1995    0.026284
2016    0.026134
2013    0.017201
2014    0.014202
1994    0.013474
1993    0.009104
2015    0.008397
1992    0.007947
1990    0.007433
1991    0.007262
1989    0.003727
1988    0.002892
1985    0.002056
          ...   
1966    0.000471
1977    0.000471
1976    0.000450
1969    0.000407
1975    0.000386
1965    0.000364
1964    0.000257
1963    0.000171
1961    0.000129
1959    0.000129
1956    0.000086
1958    0.000086
1937    0.000086
1962    0.000086
1950    0.000064
1954    0.000043
1941    0.000043
1934    0.000043
1957    0.000043
1951    0.000043
1955    0.000043
1931    0.000021
1953    0.000021
1943    0.000021
1938    0.000021
1939    0.000021
1927    0.000021
1929    0.0000

Looks like most cars are registered in the past 10 years. Cars with model year 2000 was number 1.

## Exploring Price by Brand

In [30]:
brand_counts = autos["brand"].value_counts(normalize=True)
top_brands = brand_counts[brand_counts > 0.01].index
print(top_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai'],
      dtype='object')


German brands take the top 5, takes more than 50% of total brands listed. Let's choose brand with at least 1%.

In [33]:
brand_mean_prices = {}

for brand in top_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices

{'audi': 9336,
 'bmw': 8571,
 'citroen': 3779,
 'fiat': 2813,
 'ford': 4054,
 'hyundai': 5365,
 'mazda': 4112,
 'mercedes_benz': 8628,
 'nissan': 4743,
 'opel': 2976,
 'peugeot': 3094,
 'renault': 2475,
 'seat': 4397,
 'skoda': 6368,
 'smart': 3580,
 'toyota': 5167,
 'volkswagen': 5604}

Highest prices go to audi, bmw, and mercedes_benz, lowest prices to fiat, renault, and opel

## Storing Aggregate Data in a DataFrame

In [35]:
bmp_series = pd.Series(brand_mean_prices)
pd.DataFrame(bmp_series, columns=["mean_mileage"])


brand_mean_mileage = {}

for brand in top_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)

brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info

mean_mileage
bmw                  132575
mercedes_benz        130788
opel                 129311
audi                 129157
volkswagen           128711
renault              128127
peugeot              127153
mazda                124464
ford                 124266
seat                 121131
citroen              119694
nissan               118330
fiat                 117121
toyota               115944
skoda                110848
hyundai              106442
smart                 99326

In [36]:

brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
bmw                  132575        8571
mercedes_benz        130788        8628
opel                 129311        2976
audi                 129157        9336
volkswagen           128711        5604
renault              128127        2475
peugeot              127153        3094
mazda                124464        4112
ford                 124266        4054
seat                 121131        4397
citroen              119694        3779
nissan               118330        4743
fiat                 117121        2813
toyota               115944        5167
skoda                110848        6368
hyundai              106442        5365
smart                 99326        3580


The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.